### DISCUSSION TITLE
Count Dice equal to the dice target, and increase dice up to the dice target with total of Roll points mechanic help.

### DISCUSSION THREAD

**Discussion by wmrobmuadib** [posita/dyce#11] [Q&A] [answered]
Posita, and anyone else able to help. I am still a python noob, so finding this one hard to put together. 

My mechanic has you roll a pool (based on your 1-20 attribute (10 average) plus dicepool modifiers) of 20 sided dice looking to meet a dice target, 20 by default, but it varies (from 2- 20+-30-40 etc.). Each die that meets or beats the die target counts as 1 result point. Now, dice targets can increase above 20. but, if a die comes up a natural 20, it explodes and you reroll and add to see if you can meet the greater than 20 dice target. 

So that's not too complicated to figure out. BUt here is where I get lost. You also have a number of roll points equal to your skill. After rolling and rerolling any nat 20s. you can add these roll points to individual die totals to reach the dice target. Each one boosted that way counts as an additional *result* point. If you don't have enough roll points to boost a die, or enough remaining roll points, they are ignored. 

natural 1s can't have dice points added to them. Additionally, if none of the dice in your dicepool reach the dice target, it counts as 0 result points, a failure. Additionally, if none of the dice meet the dice target, each natural 1 in the dicepool counts -1 result point, a failure result point. 

Now I am confused as how to handle all this, being a python novice. If you could code this out and explain it to me, I would appreciate it. 

Thanks.

**Comment by posita** [comment]
Thanks for the question @wmrobmuadib! Before I offer a solution, I'd like to understand the mechanic a little better.

How to roll points work? Is it a fixed bonus that is added to each d20 rolled, or is it a single amount that the player can divide and apply to zero or more of the values rolled? If the latter, is it assumed that the player would always optimally allocate the points to maximize the number of result points?

Do natural 1s affect the aggregate result points *only* if no value in the roll meets or beats the target number? For example, let's say the target number is 10, and my pool size is 5, and I have no roll points. I roll: 1, 1, 2, 2, 12. Is the result 1 or -1?

**Comment by wmrobmuadib** [comment]
THanks for your response!

ROll points are a limited pool that can be individually spent to add to a specific die total to bring it up to the die target. And the rule is you apply the roll points to the highest die less than the dice target to bring it to the dice target. If there are roll points left over after adding to that die, you add points to the next highest die, if you don't have enough roll points remaining to add to it so that it reaches the die total, the remaining roll points are lost/ignored. 

And natural ones only affect result points if no value in the roll meets or beats the target number, yes.  In your example, the result 1, because the 12 is greater than or equal to the dice target. If it was a 9, then you would look for 1s for negative result points. 

Thanks for your help.

**Comment by posita** [comment, marked as answer]
Hi @wmrobmuadib! It took me awhile to get to this, but I think I have you covered. The bad news is that, with so many D20s involved, your mechanic requires some computational heavy lifting (at least with respect to `dyce` and other traditional means of enumeration). The good news is that @HighDiceRoller's [Icepool](/HighDiceRoller/icepool) provides optimizations that help out **_a lot_**.

I have a write-up here: 👉👉👉 [![Try ``dyce``](https://jupyterlite.readthedocs.io/en/latest/_static/badge.svg)](https://posita.github.io/dyce-notebooks/lab?path=github%2Froll-points-posita-dyce-11%2Froll_points.ipynb) 👈👈👈 *[[source](https://github.com/posita/dyce-notebooks/tree/main/notebooks/github/roll-points-posita-dyce-11)]*

The even better news is that translating between `dyce` and Icepool primitives is straightforward and allows us to use [`anydyce`](/posita/anydyce)'s visualization conveniences. I hacked together a small interactive widget that allows you to explore your mechanic in that write-up:

[![Screenshot from 2023-03-31 16-37-52](https://user-images.githubusercontent.com/222581/229238380-faf61e6f-fefc-4503-a771-d899ba40035d.png)](https://posita.github.io/dyce-notebooks/lab?path=github%2Froll-points-posita-dyce-11%2Froll_points.ipynb)

**Reply by wmrobmuadib** [reply]
Thanks so much! And wow Icepool is pretty impressive for handing the really computionally heavy stuff.  Can't say I understand it under the hood much. But thanks again for your hard work. This gives me a good feel for how the mechanic shakes out. And the widget is very handy. I  tried some traditional enumeration and it about crashed my computer for a large pool. The optimized methods are quite efficient in comparison.

**Reply by HighDiceRoller** [reply]
Thanks for the mention and interest! I think @posita pretty much covered it; the only other thing I have to add is a link to the [previous time I saw a similar system to your roll points](https://math.stackexchange.com/questions/4403443/probability-of-rolling-n-dice-that-are-each-are-greater-than-or-equal-to-x-w) (though essentially *only* roll points).

**Reply by wmrobmuadib** [reply]
Interesting, I saw the roll points idea mostly from Mythic D6, an evolution of the DC Universe RPG< which was a simplified WEG D6 version. Where they changed from totaling the dice to counting 'successess/hits" for dice results. Mythic D6 added back pips which let you alter your highest dice like in my spending system. It also has a wild die that explodes, but only a 6, which spawns additional successes if you roll a 4,5 and an additional success plus roll again on a 6. I actually added a dual Natural 20 rule which I have evaluated.    (THis is gonna get long)  Basically it is a 6 step process to generate the outcome points for a check. a couple of which are optional/uncommon. 


STEP 1: Gather your DIcepool  Make INitial ROll
FIrst, is you gather your dicepool of 20 sided dice, the number of which is equal to your acting attribute, plus any dicepool modifiers for internal circumstances, i.e. penalties for injury, fatigue, disease, encumbrance, and bonuses for taking extra time, good equipment etc.   attributes range from 0 to 20 for human, up to 30 for extrahuman. There is a 20 dice rule, if you would roll more than 20 dice after modifiers and such, you only roll 20 dice, and add roll points to the roll equal to the excess dice over 20. So a 30 dice pool would be 20 dice +10 roll points modifier.  you then roll your initial dicepool

STEP 2: Apply Action Bias
After rolling your initial dicepool. you apply any Action BIas, which reperesents metaphysical intertia that exists in some Spheres (My game takes place in a multiverse with hardcoded reality rules). Such that positive action bias makes certain favored actions easier, and negative action bias makes certain actions harder.  THis is accomplished by a reroll rule.   By default, you don't reroll any of your initial dice.   THe reroll map is   

# Reroll conditions for action biases
ACTION_BIAS_REROLL_CONDITIONS = {
    "Ordained": (1, 2, 3),
    "Catalyzed": (1, 2),
    "Favored": (1,),
    "Neutral": (),
    "Disfavored": (20,),
    "Inhibited": (19, 20),
    "Proscribed": (18, 19, 20),
}

So if your attempting something that is disfavored, you have to reroll once, and only once, high rolling dice. Making it less likely you will succeed. while positive bias actions are the opposite. 

STEP 3 - Roll and Add Bonus dice. 
I didn't include this mechanic option in my writeup for Posita. So It isn't accounted for in the code he graciously put together for me.   So the rule is, for any Natural 20s, in your dicepool. after applying action bias. You gather another die, called a bonus die, and roll and add it to your dicepool. These additional bonus dice, can spawn bonus dice and so on. 

Step 4 - Explode Natural 20s 
If the DIce Target is greater than 20, the default. Here is another rule that is accounted for. Dice targets for action checks are 20 by default. However, Target Modifiers can change this, as can Differences is Ability CLass So someone with Superhuman +3 ability versus someone with Normal +0 ability would have a base dice target of (20-(3-0)=17) To which target modifiers for external circumstances would be added up, and the cumulative modifier applied.    Which brings us back to natural 20s. If the dice target is greater than 20. Any natural 20 explodes. Reroll the 20 sided die and add the number to the intital 20 to get the exploded die. At this point you have your Post-Action BIas, Post BOnus Dice, Post _explosion dice pool. 

STEP 5 - Apply roll points. 
Characters have a limited pool of roll points equal to their skill level which can be spent to improve dice totals on each check. To spend these, you follow a simple algorithm. First look for the highest dice LESS than the dice target, that aren't natural 1s. (Natural 1s are auto failures, and can't have roll points added to them, no matter how many you have.)  If you have enough points to raise it to the dice target, spend those points and raise the die to the dice target. If you don't have enough, then you ignore the roll points and move to step 6. If you have enough, and have additional roll points remaining, You look for the next highest die less than the dice target, and determine if you have enough roll points to add to that roll to reach the dice target. If so you spend them, and move on to the next die, etc. So long as you have roll points remaining and they are enough to raise the roll to the dice target. Otherwise, you ignore the remaining roll points and move to Step 6.

STEP 6 - Count Outcome Points 
For each die whose post action bias, post bonus die, post explosion, post roll point individual die total is greater than or equal to the dice target, you score one Outcome point. If none of your modified individual die totals equal or exceed the the dice target, the roll is scored as 0 outcome points, a failure. . Further, for each natural 1 in the dicepool. you accrue -1 outcome points, a seriously negative outcome. 

Interestingly enough, as I am still a python n00b. I cobbled together a monte carlo sim functional version with the help of GPT4. I have attached it here if your want to look. It is not nearly as elegant as using the Dyce and Icepool objects. BUt the numbers appear very close to the enumerated results the script put together by Posita generated.  It is in plain text, but it is a jupyter notebook file .ipynb




[D20ExNBNsPlusRPVsTN-RollerNOddsWActionBIas.txt](https://github.com/posita/dyce/files/11128749/D20ExNBNsPlusRPVsTN-RollerNOddsWActionBIas.txt)
